pymatviz/ptable.py

```diff
@@ -160,6 +214,418 @@ def count_elements(
+                        raise TypeError(f"Unsupported data type {type(value)}") from exc
+
+        # Set vmin and vmax
+        df.attrs["vmin"] = min(flattened_values)
+        df.attrs["vmax"] = max(flattened_values)
+
+        return df
+
+    # Check and handle different supported data types
+    if isinstance(data, pd.DataFrame):
+        data_df = data
+
+    elif isinstance(data, pd.Series):
+        data_df = data.to_frame(name="Value")
+        data_df.index.name = "Element"
+
+    elif isinstance(data, dict):
+        data_df = pd.DataFrame(data.items(), columns=["Element", "Value"]).set_index(
+            "Element"
+        )
+
+    else:
+        raise TypeError(f"Unsupported data type, choose from: {SupportedDataType}.")
+
+    # Convert all values to np.array
+    data_df["Value"] = data_df["Value"].map(
+        lambda x: np.array([x]) if isinstance(x, float) else np.array(x)
+    )
+
+    # Handle missing and anomaly
+    data_df = handle_missing_and_anomaly(data_df)
+
+    # Write vmin/vmax into metadata
+    return set_vmin_vmax(data_df)
+
+
+def handle_missing_and_anomaly(
+    df: pd.DataFrame,
+    # missing_strategy: Literal["zero", "mean"] = "mean",
+) -> pd.DataFrame:
+    """Handle missing value (NaN) and anomaly (infinity).
+
+    Infinity would be replaced by vmax(∞) or vmin(-∞).
+    Missing values would be handled by selected strategy:
+        - zero: impute with zeros
+        - mean: impute with mean value
+
+    TODO: finish this function
+    """
+    return df
+
+
+class PTableProjector:
+    """Project (nest) a custom plot into a periodic table.
+
+    Scopes mentioned in this plotter:
+        plot: Refers to the global scope.
+        ax: Refers to the axis where child plotter would plot.
+        child: Refers to the child plotter itself, for example, ax.plot().
+    """
+
+    def __init__(
+        self,
+        data: SupportedDataType,
+        colormap: str | Colormap | None,
+        plot_kwargs: dict[str, Any] | None = None,
+    ) -> None:
+        """Initialize a ptable projector.
+
+        Default figsize is set to (0.75 * n_groups, 0.75 * n_periods),
+        and axes would be turned off by default.
+
+        Args:
+            data (SupportedDataType): The data to be visualized.
+            colormap (str | Colormap | None): The colormap to use.
+            plot_kwargs (dict): Additional keyword arguments to
+                pass to the plt.subplots function call.
+        """
+        # Get colormap
+        self.cmap: Colormap = colormap
+
+        # Preprocess data
+        self.data: pd.DataFrame = data
+
+        # Initialize periodic table canvas
+        n_periods = df_ptable.row.max()
+        n_groups = df_ptable.column.max()
+
+        # Set figure size
+        plot_kwargs = plot_kwargs or {}
+        plot_kwargs.setdefault("figsize", (0.75 * n_groups, 0.75 * n_periods))
+
+        self.fig, self.axes = plt.subplots(n_periods, n_groups, **plot_kwargs)
```

Comment:
> I think both should work almost the same because in my class, there isn't much (if any) shared states at all.

i would consider self.(...) attribute as state. but rather than belabor the point, let's merge this PR as is nad refactor later if needed.